### PR TITLE
Ele sf

### PR DESCRIPTION
**Pull request by ahmad3213**

### CODE REVIEW

File: corrections.h [hunk 1]
```diff
@@ -13,6 +13,7 @@ using RVecF = ROOT::VecOps::RVec<float>;
 using RVecB = ROOT::VecOps::RVec<bool>;
 using RVecVecI = ROOT::VecOps::RVec<RVecI>;
 using RVecLV = ROOT::VecOps::RVec<LorentzVectorM>;
+using RVecUL = ROOT::VecOps::RVec<unsigned long>;
```
Comment (kandrosov) [inline corrections.h:16]:
```suggestion
```

File: electron.h [hunk 2]
```diff
@@ -41,43 +41,39 @@ class EleCorrProvider : public CorrectionsBase<EleCorrProvider> {
...
+        RVecLV final_p4 = Electron_p4;
+        for(size_t n = 0; n < Electron_p4.size(); ++n) {
+            const GenLeptonMatch genMatch = static_cast<GenLeptonMatch>(Electron_genMatch.at(n));
+            double sf = 1 ;
+             if(genMatch==GenLeptonMatch::Electron || genMatch == GenLeptonMatch::TauElectron)
+             {
+            sf = EleES_->evaluate({"total_uncertainty",static_cast<int>(Electron_seedGain.at(n)), static_cast<double>(run), Electron_p4[n].eta(), static_cast<double>(Electron_r9.at(n)), Electron_p4[n].pt()});
+            if (getESScaleStr(scale) == "scaleup") final_p4[n] *= (1+sf);
+            if (getESScaleStr(scale) == "scaledown") final_p4[n] *= (1-sf);
+              }
```
Comment (kandrosov) [inline electron.h:72]:
```suggestion
            if(scale != UncScale::Central
                    && (genMatch==GenLeptonMatch::Electron || genMatch == GenLeptonMatch::TauElectron) ) {
                sf = EleES_->evaluate({"total_uncertainty", static_cast<int>(Electron_seedGain.at(n)), run,
                                       Electron_p4[n].eta(), static_cast<double>(Electron_r9.at(n)),
                                       Electron_p4[n].pt()});
                final_p4[n] *= 1 + static_cast<int>(scale) * sf;
            }
```

File: electron.py [hunk 3]
```diff
@@ -14,24 +14,45 @@
...
     ID_sources = ["EleID"]
     working_points = ["wp80iso", "wp80noiso"]
+    energyScaleSources_ele = ["EleES"]
     year = ""
 
     def __init__(self, period):
         EleID_JsonFile = EleCorrProducer.EleID_JsonPath.format(period)
-        EleES_JsonFile = os.path.join(os.environ['ANALYSIS_PATH'],EleCorrProducer.EleES_JsonPath.format(period))
+
+        if ( ('2016' in period) or ('2018' in period) or ('2018' in period)):
```
Comment (kandrosov) [inline electron.py:27]:
```suggestion
        if period.startswith('Run2'):
```

File: electron.py [hunk 4]
```diff
@@ -14,24 +14,45 @@
...
 
         if not EleCorrProducer.initialized:
             headers_dir = os.path.dirname(os.path.abspath(__file__))
             header_path = os.path.join(headers_dir, "electron.h")
             ROOT.gInterpreter.Declare(f'#include "{header_path}"')
-            ROOT.gInterpreter.ProcessLine(f'::correction::EleCorrProvider::Initialize("{EleID_JsonFile}", "{EleES_JsonFile}")')
+            ROOT.gInterpreter.ProcessLine(f'::correction::EleCorrProvider::Initialize("{EleID_JsonFile}", "{EleES_JsonFile}","{EleID_JsonFile_key}","{EleES_JsonFile_key}")')
             EleCorrProducer.year = period.split("_")[0]
+            if (period.endswith('Summer22')):  EleCorrProducer.year = period.split("_")[0]+"Re-recoBCD"
+            if (period.endswith('Summer22EE')):  EleCorrProducer.year = period.split("_")[0]+"Re-recoE+PromptFG"
```
Comment (kandrosov) [inline electron.py:44]: add 2023 eras

File: met.py [hunk 5]
```diff
@@ -30,7 +30,7 @@ def getPFMET(self, df, source_dict):
                     p4_delta_list = [ f'{obj}_p4_{syst_name}_delta' for obj in source_objs ]
                     p4_delta_str = ', '.join(p4_delta_list)
                     df = df.Define(f'MET_p4_{syst_name}',
-                                   f'::correction::ShiftMet(MET_p4_{nano}, {{ {p4_original_str} }}, {{ {p4_shifted_str} }}, false)')
-                    df = df.Define(f'MET_p4_{syst_name}_delta', f'MET_p4_{syst_name} - MET_p4_{nano}')
+                                   f'::correction::ShiftMet(PFMET_p4_{nano}, {{ {p4_original_str} }}, {{ {p4_shifted_str} }}, false)')
+                    df = df.Define(f'MET_p4_{syst_name}_delta', f'MET_p4_{syst_name} - PFMET_p4_{nano}')
```
Comment (kandrosov) [inline met.py:34]: take MET name from the config

**Review by kandrosov — CHANGES_REQUESTED** [review]
also resolve merge conflicts

**Comment by ahmad3213** [comment]
> add single electron trigger SF to this PR (same as it was for muons)

This will come later, next week. it will touch different files, so if you can merge this before please

File: electron.h [hunk 6]
```diff
@@ -41,44 +41,42 @@ class EleCorrProvider : public CorrectionsBase<EleCorrProvider> {
...
         correctionsES_(CorrectionSet::from_file(EleESFile)),
-        EleIDSF_(corrections_->at("UL-Electron-ID-SF")),
-        EleES_(correctionsES_->at("UL-EGM_ScaleUnc"))
+        EleIDSF_(corrections_->at(EleIDFile_key)),
+        EleES_(correctionsES_->at(EleESFile_key))
+        {
+        }
+
+    float getID_SF(const LorentzVectorM& Electron_p4, std::string working_point, std::string period, UncSource source, UncScale scale) const
     {
```
Comment (kandrosov) [inline electron.h:53]:
```suggestion
```

**Review by kandrosov — CHANGES_REQUESTED** [review]

**Review by kandrosov — APPROVED** [review]

**Pull request merged**